Aggregated Mail Notifications

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/20694
## Description

This PR aggregates the mail notifications a user receives for work packages according to the instance wide settings.

That means:
- Mail delivery is delayed by the configured amount
- the delivered mail should include the latest aggregated journal
- for each aggregated journal only a single mail is delivered
  - e.g. for two subsequent (quick) updates only one mail is delivered
## Notes

(CC @linki [source](https://github.com/finnlabs/openproject-multitenancy/blob/dev/config/initializers/mail_notification_job.rb))

This PR removes the `MailNotificationJob` and its setting `raise_exceptions`. It should not be needed anymore, since we raise all exceptions that occur. For expected edge cases (e.g. the journal does not exist anymore, because the WP was deleted), we behave in an determined and exception-free manner.

NobodysNightmare committed:
send aggregated notifications

- only sending the last mail of an aggregation chain
- displaying the aggregated state of the change
- replaces old classes for mail notifications
- old special error handling for missing users is obsoleted, since recipients are determined at jobs runtime
- special cases for missing author and journal are considered

NobodysNightmare committed:
spec more cases

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-aggregated-mails.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-aggregated-mails.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
do not let side-effects happen on passed parameter

NobodysNightmare committed:
specify fancy edge-case

yet unsolved

NobodysNightmare committed:
partly address an edge case that can prevent notification delivery

"partly" because the current implementation is still "vulnerable" if the succeeding journal gets outdated in the same way as the preceeding one. The edge-case of the edge-case....

NobodysNightmare committed:
split up mail delivery and enqueueing

NobodysNightmare committed:
correctly disable notifications during spec

the way used earlier does unregister notifications for the rest of the test run

NobodysNightmare committed:
move on-behalf-delivery into notification mailer

oliverguenther commented on app/models/user.rb:
```diff
@@ -716,6 +716,14 @@ def self.current
     @current_user ||= User.anonymous
   end
 
+  def self.execute_as(user)
```
While I see that this existed before this PR and thus may be out of scope, I really dislike this approach. Why exactly is this override necessary? I couldn't detect anything in the used UserMailer methods. Can you provide me with some insights as to why this remains necessary?

NobodysNightmare replied:
E.g. here: https://github.com/opf/openproject/blob/108160ed88dc1d8e97e21d141f1d0a668c68a5d8/app/views/user_mailer/work_package_updated.html.erb#L35

The problem is that we parse the textile for mails too. And parsing textile requires the `User.current` to be set correctly (e.g. "does the user have the permission to expand the macro #123?").

The _good way™_ would be to allow injecting a user into the text formatter. But I deemed that out-of-scope at the time I wrote the fix (greatly because of feasibility/time constraints).

oliverguenther replied:
Ah, thanks. I missed that skimming through the templates. Makes sense, but if there's any time (left), we should strive to modify this.

NobodysNightmare committed:
add specs for notification mailer part of edge case

NobodysNightmare commented:
I am starting to hate this PR. It started quite nice and then the edge cases arrived and did not stop.

I have the feeling I should try to DRY it up some more. On the other hand it already took an eternity until this point and we only have so much time for development.

I am opening this for review now. Please somebody tell me that I should go and fix it ^^

NobodysNightmare commented:
Found a new issue -.-
- [ ] changing an attribute twice results in "on behalf" delivery

oliverguenther commented:
I can't seem to reproduce this last one, or misunderstood the issue you found:

I changed the same attribute (priority from Normal to High, then High to Immediate) within one aggregation interval and am receiving

> Work package #45 has been updated by OpenProject Admin.
> Priority changed from Normal to Immediate

That is the correct author, and the correct end result. (?)

NobodysNightmare commented:
@oliverguenther I found a solution with @ulferts for the problem. However I have to admit that I failed to put the reproduction into a simple test case until now.
My tests always worked on the "% done", which I changed twice. My element was a child element of another WP (if that makes any difference).

I'll try to reproduce the exact circumstances to put them into a test case.

oliverguenther commented:
Also happens when creating a new work package in the hierarchy. The child itself is aggregated correctly (`Work package #78 has been reported by OpenProject Admin. ... correct last state of attributes`, but all intermittent changes are reported individually on the parent.

NobodysNightmare committed:
fix wrongful notificaiton delivery

NobodysNightmare committed:
fix specs

NobodysNightmare commented on spec/legacy/unit/journal_spec.rb:
```diff
@@ -37,21 +37,19 @@
       issue.save
     end
-    user = User.find(:first)
-    assert_equal 0, ActionMailer::Base.deliveries.size
```
I don't know where this assertion stemmed from.

How can I expect no notifications **after** adding a journal?

I made the test pass, but I changed the expectation quite harshly, so please someone point out if there was a reason for the original assertion, that I don't see...

oliverguenther replied:
I've digged through that particular spec a bit (even though my previous knowledge of the journal is limited). What I've found was that on `release/4.3`, `Setting.notified_events.include?('work_package_added')` was set to false, thus the whole notification in `handle_work_package_create` was not evaluated, and no deliveries were made.

Recall the original structure of `handle_work_package_create`: 

``` ruby
def handle_work_package_create(work_package)
  if Setting.notified_events.include?('work_package_added')
    # queue job
  end
end
```

Thus the assertion would be valid, as no deliveries should have been made

In your changes, `send_notification?` evaluates to true due to
- `(Setting.notified_events.include?('work_package_added') && journal.initial?)` being false (the former unchanged being false as in release/4.3)
- `Setting.notified_events.include?('work_package_updated')` **being true**

I'm not sure if this is a big issue in the updated code, but I guess you'd have a better insight here to tell if that makes a significant difference.

NobodysNightmare replied:
Narf!

Of course that makes a difference! It is a shame that the specs for the notification mailer did not unveil it. Thanks for pointing out.

The difference here is that I will now send a notification based on the **update** setting in the **create** case. I should not be doing that.

oliverguenther replied:
It didn't sound right upon discovery, but with all your opaque black magic going on here I was rather careful with my conclusions ^^

oliverguenther commented:
Travis failure due to (probably) unrelated feature spec on pg only, restarted

NobodysNightmare committed:
fix bad behaving notifications

NobodysNightmare committed:
add test case for hard edge-case of edge-case

oliverguenther commented:
The on behalf of notification issue seems to be resolved. Feature LGTM :+1:
